### PR TITLE
[DMS-277] DB schema changes for delete performance

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/DeleteTests.cs [hunk 1]
```diff
@@ -122,6 +122,7 @@ public void It_should_be_a_successful_delete_for_1st_transaction()
         }
 
         [Test]
+        [Ignore("DMS-296 will resolve intermittent serialization failures causing UnknownFailure")]
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/DeleteTests.cs:125]: See DMS-296, but ran into this several times. Not sure why.

File: src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/UpdateTests.cs [hunk 2]
```diff
@@ -235,76 +236,63 @@ public async Task It_should_be_the_1st_update_found_by_get()
     }
 
     [TestFixture]
-    public class Given_an_update_of_a_document_that_references_a_non_existent_document : UpdateTests
+    public class Given_an_update_of_a_document_to_reference_a_non_existent_document : UpdateTests
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/UpdateTests.cs:239]: This test was not testing the right thing

File: src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/UpdateTests.cs [hunk 3]
```diff
@@ -385,66 +365,58 @@ public class Given_an_update_of_a_document_with_one_existing_and_one_non_existen
...
-        {
-            _upsertResults.Should().HaveCount(2);
-            _upsertResults.ForEach(x => x.Should().BeOfType<UpsertResult.InsertSuccess>());
-        }
-
         [Test]
         public void It_should_be_a_update_failure_reference()
         {
-            _updateResult.Should().BeOfType<UpdateResult.UpdateFailureImmutableIdentity>();
+            _updateResult.Should().BeOfType<UpdateResult.UpdateFailureReference>();
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/UpdateTests.cs:411]: This test was not testing the right thing

File: src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/UpdateTests.cs [hunk 4]
```diff
@@ -588,19 +551,20 @@ public void It_should_be_a_successful_update_for_1st_transaction()
...
             _updateResult2!.Should().BeOfType<UpdateResult.UpdateFailureWriteConflict>();
         }
     }
 
     // Future tests - from Meadowlark
-    
-    // given an update of a document that references an existing descriptor
 
-    // given an update of a document that references a nonexisting descriptor
+    // given an update of a document that tries to reference an existing descriptor
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/UpdateTests.cs:563]: Rename to make scenario more clear

File: src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs [hunk 5]
```diff
@@ -13,130 +13,18 @@
 
 namespace EdFi.DataManagementService.Backend.Postgresql.Operation;
 
-/// <summary>
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs:16]: This unnecessary duplication was getting to be too much.

File: src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs [hunk 6]
```diff
@@ -177,6 +65,7 @@ LockOption lockOption
                 }
             };
 
+        await command.PrepareAsync();
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs:68]: All SQL now cached by PostgreSQL as prepared statements

File: src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs [hunk 7]
```diff
@@ -480,9 +376,28 @@ NpgsqlTransaction transaction
...
+                  ids.documentId,
+                  ids.documentPartitionKey,
+                  ids.referentialId,
+                  ids.referentialPartitionKey,
+                  a.documentId,
+                  a.documentPartitionKey
+              FROM unnest($1, $2, $3, $4) AS
+                  ids(documentId, documentPartitionKey, referentialId, referentialPartitionKey)
+              LEFT JOIN dms.Alias a ON
+                  ids.referentialId = a.referentialId and ids.referentialPartitionKey = a.referentialPartitionKey",
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs:400]: Inserts into references table now requires a join to get documentId of reference.

File: src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs [hunk 8]
```diff
@@ -611,12 +530,14 @@ LockOption lockOption
...
-                   WHERE d2.DocumentUuid =$1 AND d2.DocumentPartitionKey = $2) AS re
-                   ON re.ParentDocumentId = d.id AND re.ParentDocumentPartitionKey = d.DocumentPartitionKey
+                $@"SELECT d.ResourceName FROM dms.Document d
+                   INNER JOIN (
+                     SELECT ParentDocumentId, ParentDocumentPartitionKey
+                     FROM dms.Reference r
+                     INNER JOIN dms.Document d2 ON d2.Id = r.ReferencedDocumentId
+                       AND d2.DocumentPartitionKey = r.ReferencedDocumentPartitionKey
+                       WHERE d2.DocumentUuid = $1 AND d2.DocumentPartitionKey = $2) AS re
+                     ON re.ParentDocumentId = d.id AND re.ParentDocumentPartitionKey = d.DocumentPartitionKey
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs:540]: One less join now to get post-delete failure info. All joins are also now on indexes.

File: tests/RestClient/demo.http [hunk 9]
```diff
@@ -0,0 +1,137 @@
+@port = 5198
+
+### Note, no descriptor validation yet
+
+### Setup - POST School Year 2025
+POST http://localhost:{{port}}/data/ed-fi/schoolYearTypes
```
Comment (bradbanister) [inline tests/RestClient/demo.http:6]: Friday's demo

Comment (stephenfuqua) [reply tests/RestClient/demo.http:6]: This will be great for the Project Tanager workgroup meeting on Thursday.

**Review by stephenfuqua — APPROVED** [review]

**Pull request merged**